### PR TITLE
adding BP005 as draft: new autils repository

### PR DESCRIPTION
**Pull request by beraldoleal**
Signed-off-by: Beraldo Leal <bleal@redhat.com>

### CODE REVIEW

**Review by clebergnu — COMMENTED** [review]
Hi @beraldoleal, thanks for posting this draft!  As one of the authors, I guess I don't have much to *review*, but I did have a few initial comments.

It would be very valuable for all of the Avocado (and associated projects like Avocado-VT and avocado-misc-tests) to engage in this discussion.  I've initially tagged @luckyh @dzhengfy @PraveenPenguin @Naresh-ibm @pevogam but this is just a way to start the reverberation of the discussion.  Please invite other contributors!

**Comment by luckyh** [comment]
> It would be very valuable for all of the Avocado (and associated projects like Avocado-VT and avocado-misc-tests) to engage in this discussion. I've initially tagged @luckyh @dzhengfy @PraveenPenguin @Naresh-ibm @pevogam but this is just a way to start the reverberation of the discussion. Please invite other contributors!

\+ @YongxueHong @vivianQizhu

File: docs/source/blueprints/BP005.rst [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+The exact format of the metadata file is still to be defined, but a possibility
+is to define a JSON schema that will be provided on this new repository.
+
+Maintainability
+===============
+
+As described above, each library must follow the new standard and keep that
+same level for future versions. Because of that, new PRs must be approved by at
+least one MAINTAINER and pass the basic checks pipelines that will be provided
```
[the remark this answers is not in the document]
Comment (Naresh-ibm) [reply docs/source/blueprints/BP005.rst:171]: @beraldoleal @clebergnu module wise MAINTAINER/s is really a good thought.
want to hear more on roles and authorities of MAINTAINER/s here. 
is this the same what we have currently or added with any axtra responsibilities.

File: docs/source/blueprints/BP005.rst [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+Every module will need to provide a description including:
+
+* Utility description
+* The utility category (ie: network, cpu, security, etc..)
+* Maintainers information
+  * Main contact (We need a contact point, a single maintainer or a group)
+  * Maintainer Group members (this will be used to block merges)
+* Supported Platforms
+* Supported execution method (Local vs Remote)
```
Comment (pevogam) [inline docs/source/blueprints/BP005.rst:154]: What is meant by this? Is this somehow related to `aexpect`?

File: docs/source/blueprints/BP005.rst [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+For instance, the textual category and description will allow users to search
+for modules based on their interest.  Also, our pipelines will parse that data
+field containing the list of supported platforms, and will trigger tests on
+those platforms. Users will then be able to see the results of those tests on
+each of the platforms defined, giving them quality assurance and coverage
```
[the remark this answers is not in the document]
Comment (pevogam) [reply docs/source/blueprints/BP005.rst:160]: Not directly related to the search comments here but I hope you notice how your variant definitions start looking more and more like Avocado VT-s, the only difference being that instead we use containers on Github :o)

File: docs/source/blueprints/BP005.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (pevogam) [reply docs/source/blueprints/BP005.rst:171]: Indeed, this is where Avocado VT has tried to be for years now but never went formally. At the moment lots of different areas of code there have 1 or less maintainers that everybody pings when respective changes are introduced by a PR and we all do this implicitly just from memory. There is a MAINTAINERS file there but it points to general maintainers without any finer resolution on whom to ping about a particular change. In this regard I am happy to see this repo and blueprint trying to amend what has been a long-standing problem there and one that could often freeze a lot of PR-s indefinitely and discourage contribution of any kind.

In summary: I really like this idea and it scales a lot better for larger teams, of course assuming that there is sufficiently strong automation to iron out most of the difference in style and approach among contributors and maintainers.

File: docs/source/blueprints/BP005.rst [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+Maintainability
+===============
+
+As described above, each library must follow the new standard and keep that
+same level for future versions. Because of that, new PRs must be approved by at
+least one MAINTAINER and pass the basic checks pipelines that will be provided
+as part of the autils repository.
+
+Besides the automatic checks, each MAINTAINER is responsible for keeping the
+consistency of that library with the overall project.
```
Comment (pevogam) [inline docs/source/blueprints/BP005.rst:175]: There will probably be validation parts that will be very hard to automate so I guess here we still ask the authors for some common sense to fill this gap but my experience in such situations has dictated that the more automation we have to replace the need of careful contributions, the better.

File: docs/source/blueprints/BP005.rst [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+
+Release
+=======
+
+As a new and standalone repository, there will not be any type of release
+version sync between Avocado and autils.  Avocado may release, say, version
+100.0, while autils won’t have a release. Because of that, autils will be
+versioned at a different pace than Avocado.
+
+Every week a automatic job will be executed to determine if there is a
```
Comment (pevogam) [inline docs/source/blueprints/BP005.rst:185]: *an automatic

File: docs/source/blueprints/BP005.rst [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+potential need for a new release:
+
+* If the number of commits since last release has reached a threshold; AND
+* There is no open discussion with the tag “release-decision”:
+  * Creates a new discussion (with the "release-decision" tag) to run a poll
+    between all the MAINTAINERS, so they will have the opportunity to thumbs
+    up if there is a need for a new release;
+  * The link to the new discussion should be sent to all MAINTAINERS;
+  * If the decision is positive, a round-robin selected maintainer should be responsible for coordinating the release it self;
```
[the remark this answers is not in the document]
Comment (pevogam) [reply docs/source/blueprints/BP005.rst:194]: Also can't the release process be automated with some trigger? E.g. continuing into a release pipeline after the polling stage? Of course I am not sure how hard this will be just from the blueprint.

File: docs/source/blueprints/BP005.rst [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+  * And the discussion should be closed
+* If there is any existing open discussion, the bot job should comment on the
+  same discussion with the updated list of commits and ping again the
+  MAINTAINERS, give them the opportunity to thumbs up based on the new
+  status;
```
Comment (pevogam) [inline docs/source/blueprints/BP005.rst:199]: Are there any other criteria for a release than just a threshold number of commits that was reached? E.g. security bug detected by GH or other bot or similar triggers?

File: docs/source/blueprints/BP005.rst [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+
+Backwards Compatibility
+***********************
+
+The current libraries are not going to be removed right away. A transition
+period is expected for migrating each library into this new standard, and
+because of that libraries in transition should be available, receive bug fixes
+for at least one LTS cycle.
```
Comment (pevogam) [inline docs/source/blueprints/BP005.rst:207]: Indeed so from this I understand that the newly contributed library at the autils repo can already receive feature improvements and similar with just stability fixes backpropagated to the source library location.

File: docs/source/blueprints/BP005.rst [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+Each library in our current code base, should receive a status flag, informing
+users what their current status is, according on the following table:
+
+1. maintainer-wanted
+
+  The library is using the legacy mode, and it is not yet converted to the
+  autils repository. This should be the status for all avocado.utils.*
+  libraries upon the creation of the autils repository.
```
Comment (pevogam) [inline docs/source/blueprints/BP005.rst:216]: How about Avocado VT's default status? Should we ping more of the Avocado VT maintainers at least to make them aware of this?

File: docs/source/blueprints/BP005.rst [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+
+2. adopted
+
+  A maintainer has picked up this library to be ported to the autils
+  repository.  It’s not expected yet that a working version is available on
+  autils, but someone is working on the migration.  During this period, the
+  avocado.utils library should continue to receive both bug fixes and
```
Comment (pevogam) [inline docs/source/blueprints/BP005.rst:223]: And once again I would probably suggest we also imply this for the Avocado VT virttest utils modules at least.

File: docs/source/blueprints/BP005.rst [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+ This means the current avocado.utils library is deprecated in favor of the
+ migrated version at the autils repository. During at least one Avocado LTS
+ cycle period, the avocado.utils version of the libraries should:
+
+  * Include warnings in each deprecated method or function, pointing users to
+    the equivalent method or function in the new autils library
+  * Keep receiving critical bug fixes backported from the new autils version
+
+ At this point, this library has an official maintainer who will be the main
+ point of contact for.
```
Comment (pevogam) [inline docs/source/blueprints/BP005.rst:239]: How about imports in general? I understand separate projects should change their imports during this period but would Avocado do so instantly? What if f.e. other projects depend on both avocado and autils? Does this imply that Avocado's usage of the migrated utils will also remains in "stable regime" for the next one year or would it start making use of new features received on the autilts side right away?

File: docs/source/blueprints/BP005.rst [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+ 
+4. abandoned
+
+  * There are no plans to migrate the library to autils.
+
+Once at least one LTS cycle has passed after the “migrated” status, the legacy
+library can be permanently removed from avocado repository.
```
Comment (pevogam) [inline docs/source/blueprints/BP005.rst:246]: Or Avocado VT repository or I guess any other repository where the library is contributed from since I didn't see a restriction requiring it to originate from `avocado.utils` only.

File: docs/source/blueprints/BP005.rst [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+**************
+
+Todo: before "approved": Create the following epic issues on GH:
+
+* Migrate one library as a "reference implementation". I.e: "network" or
+  "archive"
+* Create an improved version of our contributor guide focused on the autils
+  repository
+* Create pipeline jobs inside the new repository
+* Create release pooling bot
```
Comment (pevogam) [inline docs/source/blueprints/BP005.rst:272]: I guess you mean *polling* bot

**Review by pevogam — COMMENTED** [review]
Hi @beraldoleal, a very nice blueprint overall! Left a few comments but mostly a few questions about aspects that still seem unclear.

**Comment by dzhengfy** [comment]
> Hi @beraldoleal, thanks for posting this draft! As one of the authors, I guess I don't have much to _review_, but I did have a few initial comments.
> 
> It would be very valuable for all of the Avocado (and associated projects like Avocado-VT and avocado-misc-tests) to engage in this discussion. I've initially tagged @luckyh @dzhengfy @PraveenPenguin @Naresh-ibm @pevogam but this is just a way to start the reverberation of the discussion. Please invite other contributors!

+@chunfuwen @Yingshun @chloerh @kylazhang

File: docs/source/blueprints/BP005.rst (hunk 2 above)
Comment (clebergnu) [reply docs/source/blueprints/BP005.rst:154]: It means that we intend to flag modules that have the capability of acting on remote systems.  This will probably be a broad classification, just for informational purposes because unfortunately we have way too many ways of executing code remotely.

The `avocado.utils.network` introduced a "remote" approach based on the definition of 
a base `Host` class, along with `LocalHost` and `RemoteHost`.  Some other libraries (`avocado.utils.distro` for instance) will allow the use of a `session` parameter.  And of course `aexpect` has a lot of the same (and more) functionality.

We don't hope to address a complete standardization of the remote execution APIs just yet, but we want to at least bring some visibility to it.

File: docs/source/blueprints/BP005.rst [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+   need of having the entire Avocado framework as a dependency. This will reduce
+   the learning curve of future potential users that don't need the entire
+   framework.
+
+3. Making it clear to our users that Avocado framework is not Python-centric:
+
+   Currently, those utility libraries can only be used by Python tests and the
+   Avocado framework was designed to support tests written in multiple
+   languages.  By having those libraries outside the core harness codebase we
+   reinforce this characteristic.
```
[the remark this answers is not in the document]
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:61]: I will adapt the sentence, but I would like to keep the active voice here.

File: docs/source/blueprints/BP005.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:171]: >  is this the same what we have currently or added with any axtra responsibilities.

Probably it will have more or less what we have today, but it will be formal and well defined. Maintainers would be responsible not only of its plugin but also make sure it is following the general guide lines to make sure its plugin will not divert much from the others. Maintainers, as described here in this blueprint, will care mainly about maintainability (including quality), releases and discussions involving this new repository.

> In summary: I really like this idea and it scales a lot better for larger teams, of course assuming that there is sufficiently strong automation to iron out most of the difference in style and approach among contributors and maintainers.

Good! :) Since those plugins will be part of the same repository, they will share a common contributor guide, automation tools and pipelines. So in terms of logical, stylistic, analytical and formatter checks they could share the same rules. For sure, those rules must be agreed by us when drafting the contributor guide there.

File: docs/source/blueprints/BP005.rst [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+
+  Due to the current code quality and given this is a new repository we would
+  like to reach at least 90% of coverage. This way, users writing tests relying
+  on these libraries will have a guarantee of a good level of correctness and
+  quality, redirecting the debugging and troubleshooting efforts away from the
+  well tested libraries.
+
+The new standalone repository will have predefined pipelines to run those
+checks, so new libraries will be automatically enforced to newly introduced
+libraries.
```
[the remark this answers is not in the document]
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:141]: Thanks, fixed.

File: docs/source/blueprints/BP005.rst (hunk 2 above)
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:154]: Exactly that. Thanks @clebergnu .

File: docs/source/blueprints/BP005.rst (hunk 7 above)
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:199]: That would be a good indicator as well. Probably it will make no sense to wait for more commits if there is security bug fixed. The problem I see, is the automation here, since we don't have any errata mechanism in place (and probably don't need at this point).

So yes, we should accommodate those "manual" needs, as well. Thanks.

File: docs/source/blueprints/BP005.rst (hunk 12 above)
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:246]: I agree with the Avocado-VT part, but I guess the idea here it was to document only the parts that we control. We can't remove the library from "other repositories".

File: docs/source/blueprints/BP005.rst (hunk 13 above)
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:272]: Sure. :)

**Comment by beraldoleal** [comment]
Hi @pevogam thanks for reviewing and the comments. Just force pushed with the minor suggestions.

I'm still waiting for more comments here, so others can contribute.

File: docs/source/blueprints/BP005.rst (hunk 8 above)
Comment (clebergnu) [reply docs/source/blueprints/BP005.rst:214]: Yes, that's what is being proposed.  There is no reason for the new versions of those libraries to be bound to the capabilities of the versions under "avocado.utils".

File: docs/source/blueprints/BP005.rst (hunk 11 above)
Comment (clebergnu) [reply docs/source/blueprints/BP005.rst:246]: > How about imports in general? I understand separate projects should change their imports during this period but would Avocado do so instantly? What if f.e. other projects depend on both avocado and autils? Does this imply that Avocado's usage of the migrated utils will also remains in "stable regime" for the next one year or would it start making use of new features received on the autilts side right away?

I think each project will do an import switch at their own appropriate time.  It's a good idea indeed for Avocado to switch as soon as possible, but I'm not sure we need to make that a requirement.

For projects depending on both Avocado and autils, it should be possible to pick and choose from libraries from both places.  It'd be a bad idea for a project, maintenance and readability-wise, to use the "same" libraries at the same time from both repos.

It's my understand that Avocado would be free to start using the versions and features on the autils repo right away, even if it's keeping the old version around for some time (obvious different amount of time for LTS and non-LTS branches/versions).

File: docs/source/blueprints/BP005.rst (hunk 9 above)
Comment (pevogam) [reply docs/source/blueprints/BP005.rst:223]: Yes, my point is that the blueprint could either include mentioning of the utils there too or alternatively (and probably even better) to refer to the utils in a more generic way than the specific avocado.utils.* path. But this is all a minor thing.

File: docs/source/blueprints/BP005.rst (hunk 10 above)
Comment (pevogam) [reply docs/source/blueprints/BP005.rst:230]: Or use a more project agnostic utils wording as I mentioned above, either way it is a minor thing of course.

File: docs/source/blueprints/BP005.rst (hunk 11 above)
Comment (pevogam) [reply docs/source/blueprints/BP005.rst:246]: > For projects depending on both Avocado and autils, it should be possible to pick and choose from libraries from both places. It'd be a bad idea for a project, maintenance and readability-wise, to use the "same" libraries at the same time from both repos.

Indeed, so Avocado switching the imports and adding extra features on the autils side will usually have a motivational effect on people to also choose the new copy sooner rather than later.

> It's my understand that Avocado would be free to start using the versions and features on the autils repo right away, even if it's keeping the old version around for some time (obvious different amount of time for LTS and non-LTS branches/versions).

I see, it makes sense.

File: docs/source/blueprints/BP005.rst (hunk 12 above)
Comment (pevogam) [reply docs/source/blueprints/BP005.rst:253]: > I agree with the Avocado-VT part, but I guess the idea here it was to document only the parts that we control. We can't remove the library from "other repositories".

Yes definitely so I guess a generic referral as "...the legacy library can be permanently removed from its repository of origin" sounds clearer regarding the scope of the new repo.

**Comment by pevogam** [comment]
> Hi @pevogam thanks for reviewing and the comments. Just force pushed with the minor suggestions.
> 
> I'm still waiting for more comments here, so others can contribute.

No worries and good job @beraldoleal, I can see you put a lot of effort in this.

File: docs/source/blueprints/BP005.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (pevogam) [reply docs/source/blueprints/BP005.rst:166]: > And yes, we are now able to do the "distro" part very easily with avocado run --nrunner-spawner=podman --spawner-podman-image=$distro.

This is definitely nice but seems like we can control a single parameter in comparison to the Cartesian products of parameters and OS variations we can with Avocado VT. I actually had a vague idea to extract the Cartesian config and help our to provide it as Avocado varianter plugin if anyone is interested (not meant to simplify the learning curve unfortunately but it could provide the variant expression power many of us have grown accustomed to and can still be used with simpler configurations). If we were to migrate the main varianter for Avocado VT and most of its utilities, it will be left with just the framework of VM classes and VM isolation environments code which could be more easily digested and analyzed for potential replacement or simplification refactoring.

But I won't kidnap the topic here, my point is that we want *both* easy entry for adopters and expression power for continuous users.

> BTW, this reminds me that I owe some TLC in the LXC spawner PR (my apologies!).

No worries, I will convert it away from draft today so that it is more visible and make your lives simpler, I am aware that the draft status is not compatible with the workflow of the Avocado maintainers. I will also do my best to assist you with maintenance in the long term if you need it as this spawner will turn into core component in our Avocado testing.

**Comment by beraldoleal** [comment]
@pevogam, your suggestions where incorporated. I added a note at the top section about avocado vs avocado-vt.

Let me know if I missed something, and thanks again.

**Comment by beraldoleal** [comment]
HI @YongxueHong, @vivianQizhu, @chunfuwen, @Yingshun, @chloerh and @kylazhang: Your comments here are really valuable. Please, let us know if you have any questions.

File: docs/source/blueprints/BP005.rst [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+
+Besides the automatic checks, each MAINTAINER is responsible for keeping the
+consistency of that library with the overall project.
+
+Release
+=======
+
+As a new and standalone repository, there will not be any type of release
+version sync between Avocado and autils.  Avocado may release, say, version
+100.0, while autils won’t have a release. Because of that, autils will be
```
Comment (luckyh) [inline docs/source/blueprints/BP005.rst:192]: Here one question is: would avocado consume autils?

If yes, another question is what are the rules for avocado's plugins (for instance, avocado-vt) to consume this repo as well? Let's suppose that avocado 100.0 would have the requirement of this repo as `autils >= 5.0`, then can avocado-vt also give its own requirement (say, a different one with avocado's) to the same repo?

File: docs/source/blueprints/BP005.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (luckyh) [reply docs/source/blueprints/BP005.rst:181]: @clebergnu I'll be glad, thank you!

**Review by luckyh — COMMENTED** [review]
Hi @beraldoleal, a very good proposal! I merely have a few questions and comments as the following.

File: docs/source/blueprints/BP005.rst [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+repository on GitHub and make them also available on PyPI under `python-autils`
+package name, since the `autils` name is unavailable.
+
+Utility onboarding
+==================
+
+The primary goal of autils is to be a central place for modules that ease the
+interaction with system level features and interfaces.  As such, if a user
+needs to interact with such a system level feature or interface, he/she may
+propose a new module.
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply docs/source/blueprints/BP005.rst:100]: I see no problems with extensions in C, as long as that does not get in the way of deployment on major platforms.

I would suggest that maintainers for extensions in C have to guarantee that binary packages (autils' Python wheels) build properly and are available for users to deploy without having to have C compilers, headers, etc.

[the remark this answers is not in the document]
Comment (clebergnu) [reply docs/source/blueprints/BP005.rst:100]: That's a tough question to answer in such a generic way, because the PoV of one user may not be the same of the maintainer and vice-versa.

As a general rule, maintainers should indeed try to avoid having duplicate modules in autils if well established options are available elsewhere.  But, there may be situations in which adding a new module in autils:

* Will work better with other autils libraries (for instance, allowing remote execution transparently)
* Can simplify deployment (if autils' version has leaner requirements)
* Follow the other autils APIs, for a more consistent experience
* Will guarantee/advertise the testing of that module in different platforms

In short, I think using common sense and some differentiation checks (like the examples above) may help a maintainer to decide if a new module is worth having or not.  Of course, people proposing such a module would also need to take responsibility on its maintenance.

File: docs/source/blueprints/BP005.rst [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+
+For instance, if a user finds himself/herself constantly interacting with disk
+partitions, he/she may want to propose a “partitions” module that provides
+basic APIs for that. In this case it makes sense to have it as part of this
+central repository.
+
+However, if the utility intends to ease the interaction with an esoteric
+feature, system or device, this may not be a good place for it.
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply docs/source/blueprints/BP005.rst:108]: > I kind of understand the cases this comment against, but I just wonder if we can explain more on that. My point is we need to avoid that it would be acted by fully depending on maintainers' preference.

I believe the original meaning of this text is one in which a user writes a module that can interact with, for instance, a very specific hardware device that very few people will have access to.

For instance, let's suppose autils receives a PR like this:

---

*Subject:* Introduce autils.usrobotics

*Description:*

_This PR introduces support for manipulating switches for the Dial-Up Modems made by US Robotics in the 1990s.  Supported models include: XXXX, and YYYY._

---

The audience of the module would be extremely limited, and it would be close to impossible to have it tested on our CI.

I know the example is a bit extreme, but the same example could apply to out-of-tree, non-mainline kernel features.

But yeah, I think we can expand that topic with some examples.

File: docs/source/blueprints/BP005.rst [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+--------------------
+
+Every module will need to provide a description including:
+
+* Utility description
+* The utility category (ie: network, cpu, security, etc..)
+* Maintainers information
+  * Main contact (We need a contact point, a single maintainer or a group)
+  * Maintainer Group members (this will be used to block merges)
+* Supported Platforms
```
[the remark this answers is not in the document]
Comment (clebergnu) [reply docs/source/blueprints/BP005.rst:156]: I see no reason for excluding any platform we need to use, and we can support with the existing resources we have.

We need to keep in mind that Avocado is tested very lightly on Windows and MacOS.  Given that we want to standardize how we a test module under different platforms, I'm assuming we'll be using Avocado and appropriate spawners for different platforms.  If running under Linux and testing support for other Linux platforms, container based spawner may be enough.  If testing Windows, a QEMU/KVM spawner may be used (and developed first).

Long story short, there may be more cost involved with supporting other platforms, but there's no reason for not supporting platforms needed to fulfill our goals.

File: docs/source/blueprints/BP005.rst (hunk 16 above)
Comment (clebergnu) [reply docs/source/blueprints/BP005.rst:192]: > Here one question is: would avocado consume autils?
> 

I would expect that to be a yes.  There's a small, but maybe still meaningful, set of "avocado.utils" libraries which are used by Avocado core that *could*, with some effort, be kept as part of the Avocado repo.  But, there's a larger set of modules that are used by both Avocado core and "avocado.utils".  So, IMO, it makes sense to have all generally useful modules moved to "avocado.utils", and thus, Avocado (core) using autils.

> If yes, another question is what are the rules for avocado's plugins (for instance, avocado-vt) to consume this repo as well? Let's suppose that avocado 100.0 would have the requirement of this repo as `autils >= 5.0`, then can avocado-vt also give its own requirement (say, a different one with avocado's) to the same repo?

This is more of a Python/packaging issue than an Avocado/Avocado-VT/autils policy.  The Python solution for multiple version requirement is to have each of these combinations in a different venv.  This, of course, doesn't work when you have a layered requirement relationship (both Avocado and Avocado-VT must be on the same environment, be it a venv or not).

So, my expectation is that we would expand the idea of planning  Avocado versions known to work with Avocado-VT versions (instead of trying to guarantee compatibility at every single commit).  One of these planning points should be sync the usage of autils to the same version.

I expect that Avocado-VT, because of the requirements of newly added Avocado-VT tests (say from tp-qemu, tp-libvirt), would drive the upgrading of autils version.  Avocado (core) would then need to be adjusted to keep up the new autils changes.  I expect that the required changes on Avocado's side would be rather small, given the also small set of libraries used by Avocado core.

File: docs/source/blueprints/BP005.rst [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+  * And the discussion should be closed
+* If there is any existing open discussion, the bot job should comment on the
+  same discussion with the updated list of commits and ping again the
+  MAINTAINERS, give them the opportunity to thumbs up based on the new
+  status;
+
+Manual release process could also be started if a MAINTAINER think it is a good
+timing (i.e: Security bug or critical bug it was fixed).
+
+Backwards Compatibility
```
Comment (chunfuwen) [inline docs/source/blueprints/BP005.rst:215]: I am not sure whether it is proper to add one comment here, which is related utils API Compatibility.
In avocado-vt and tp-libvirt, those two repository heavily rely on some basic utils, such as process.run().
So it is critical to maintain those utils API Compatibility between different autils release, otherwise minor unCompatibility may lead to huge efforts change in avocado-vt and tp-libvirt.

File: docs/source/blueprints/BP005.rst [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+Besides the plan, with the migration steps for the current libraries, this
+blueprint also presents the tasks already known to be necessary in order to
+achieve this goal.
+
+
+.. note:: When reading here, in this document, "avocado.utils", please take
+          into consideration Avocado and Avocado-VT utility libraries.
+
+
+Motivation
```
Comment (chunfuwen) [inline docs/source/blueprints/BP005.rst:37]: One of part I would like to propose to add is the clear definition scope of this project. I understand it may be hard at the beginning of project. Some practices learned from successful and health upstream project ,such as kubernetes, is that those projects  define very clear project scope definition, where it is clear to point out which part should be in , which part  is not good fit-in.

File: docs/source/blueprints/BP005.rst [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+for at least one LTS cycle.
+
+Each library in our current code base, should receive a status flag, informing
+users what their current status is, according on the following table:
+
+1. maintainer-wanted
+
+  The library is using the legacy mode, and it is not yet converted to the
+  autils repository. This should be the status for all avocado.utils.*
+  (including Avocado-VT utils) libraries upon the creation of the autils
```
Comment (dzhengfy) [inline docs/source/blueprints/BP005.rst:230]: What scope of utilities is in autils as expected? Could all host related utilities be candidates? 

Are virtualization(VM) related utilities put into autils or still kept in avocado-vt? 
 
avocado-vt utils could be divided into below parts:
1. duplicated with avocado.utils.* in most extents which are not related to VM (virtualization)
2. duplicated with avocado.utils.* in some extents which are related to VM
3. consume avocado.utils.* 
4. isolated from avocado.utils.* 

For 1, move to autils
For 2, keep in avocado-vt
For 3 and 4, keep in avocado-vt

What do you think of? @beraldoleal @clebergnu

Comment (pevogam) [reply docs/source/blueprints/BP005.rst:230]: @dzhengfy What if we have nonvirtualization utils like (1) that consume avocado.utils.* like (3)? What I mean is that I don't think your separation info four category is ideal as some of the types are not mutually exclusive.

Comment (pevogam) [reply docs/source/blueprints/BP005.rst:230]: Also note that Avocado is already starting to duplicate some virtualization related functionality in the `vmimage` utility there so virtualization might actually end up needed beyond Avocado VT for spawners and other more general implementations. So I am not sure if "virtualization" will really make sense to stay in Avocado VT at least when it comes to reusable virtualization scope. Alternatively we might make sure that Avocado is always used in conjunction with Avocado VT as its main virtualization plugin.

Comment (dzhengfy) [reply docs/source/blueprints/BP005.rst:230]: > @dzhengfy What if we have nonvirtualization utils like (1) that consume avocado.utils.* like (3)? What I mean is that I don't think your separation info four category is ideal as some of the types are not mutually exclusive.

Yeah, those categories might be not very accurate as I know. The main purpose is to help clarify what the autils' scope in deed. When we are very clear about the autil's scope, then we can make the categories more accurate and decide how avocado-vt utils will go. Thanks for your caring about.

File: docs/source/blueprints/BP005.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (luckyh) [reply docs/source/blueprints/BP005.rst:100]: Sounds perfect, thanks!

File: docs/source/blueprints/BP005.rst (hunk 19 above)
[the remark this answers is not in the document]
Comment (luckyh) [reply docs/source/blueprints/BP005.rst:156]: > Long story short, there may be more cost involved with supporting other platforms, but there's no reason for not supporting platforms needed to fulfill our goals.

Yeah, I can imagine that one part of the cost would be about building up CI pipelines, which I haven't had a clear solution for that. But yes, it's good to know for having the Windows support, thank you!

File: docs/source/blueprints/BP005.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:100]: Thanks @clebergnu for clarifying.

I'm resolving this conversation so we can approve this Blueprint. Please reopen a new thread if you would like to get something fixed on this draft.

File: docs/source/blueprints/BP005.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:108]: Thanks @luckyh , I have added a new sentence at the end of that paragraph, so we can clarify this could be queried before submission in a GH thread. I will send a v2.

I'm resolving this conversation so we can approve this Blueprint.

File: docs/source/blueprints/BP005.rst (hunk 19 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:156]: Thanks @clebergnu for clarifying.

I'm resolving this conversation so we can approve this Blueprint. Please reopen a new thread if you would like to get something fixed on this draft.

File: docs/source/blueprints/BP005.rst (hunk 16 above)
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:192]: Thanks @clebergnu for clarifying.

I'm resolving this conversation so we can approve this Blueprint. Please reopen a new thread if you would like to get something fixed on this draft.

File: docs/source/blueprints/BP005.rst (hunk 20 above)
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:215]: The Backwards Compatibility section is to describe what this proposal will "break" with the current code and how.

But indeed, I agree with you @chunfuwen. Each maintainer should be responsible to make sure to use the best practices (like deprecation warnings) while updating their versions.

File: docs/source/blueprints/BP005.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:100]: Thanks @clebergnu for clarifying.

I'm resolving this conversation so we can approve this Blueprint. Please reopen a new thread if you would like to get something fixed on this draft.

File: docs/source/blueprints/BP005.rst (hunk 22 above)
Comment (beraldoleal) [reply docs/source/blueprints/BP005.rst:230]: I agree with @pevogam here. I don't see any reason for keeping virtualization libraries outside autils project. At first, we could migrate libraries that are duplicated no matter the subject (except for the cases described at the utility onboarding section).

At the same section we describe the "scope" there. It is really hard to limit the scope here at this document. Things can change quickly, so I think it might be easier for us to just review proposals to see if that makes sense or not. We may have particularities for each case.

**Comment by beraldoleal** [comment]
Closing this to send a v2. Thank you all for all the comments.
